add TF to deploy the workstation cluster

imjasonh opened the pull request:
WIP because this currently creates the cluster and workstation, but the workstation doesn't start, failing with:

> ERROR: (gcloud.beta.workstations.start) Timed out provisioning your VM because your config is degraded with message: System images cannot be pulled onto workstation VMs using this configuration. Since you have disabled public IP addresses, ensure you have enabled PGA for Artifact Registry and Container Registry by setting up DNS records for `*.pkg.dev` and `*.gcr.io` or set up Cloud NAT to allow pulling system images on your workstation VMs.

Enabling public IP addresses doesn't seem to have any effect, or I wasn't doing it right.

mattmoor commented on outputs.tf:
```diff
@@ -0,0 +1,33 @@
+ssh user@127.0.0.1 -p ${local.local_port} \
+  -o StrictHostKeyChecking=no \
+  -o UserKnownHostsFile=/dev/null
+  EOF
```
There is this command:
```
gcloud beta workstations ssh   --project=mattmoor-chainguard   --cluster=wolfi-central1   --config=wolfi-dev   --region=us-central1   wolfi-go-brrrrr
```

imjasonh replied:
Nice, I needed to update my gcloud. Thanks!

mattmoor commented on workstation.tf:
```diff
@@ -0,0 +1,61 @@
+      boot_disk_size_gb = var.disk_gb
+      // TODO: make this true and figure out how to give it access to the image.
+      disable_public_ip_addresses = false
+      // TODO: Use a less permissive service account.
+      service_account = data.google_compute_default_service_account.default.email
```
We should definitely do this, and frankly it is pointless to give it any credentials because they don't run it with `cloud-platform`, so you need to plumb auth another way.

imjasonh replied:
The service account needs access to pull the image if it's private (and I think it should be private), otherwise I'm not sure what else it needs. I tried creating a minimal SA with `roles/artifactregistry.reader` and `roles/viewer` and neither worked, and it takes foooorrrreeeeevvvver to iterate on this, so I'm sort of inclined to punt for now.

mattmoor replied:
Ah yeah, I was thinking that the GSA would be used for user auth, which is why I was disappointed about the scopes 😞

mattmoor commented on workstation.tf:
```diff
@@ -0,0 +1,61 @@
+    gce_instance {
+      machine_type      = var.machine_type
+      boot_disk_size_gb = var.disk_gb
+      // TODO: make this true and figure out how to give it access to the image.
+      disable_public_ip_addresses = false
```
Are we sure that it has access to the internet if we do that?

imjasonh replied:
I was able to ssh into the workstation at least.

This was the only way I could get this to work through the UI or TF, at least easily. We can lock it down more if we want.

mattmoor commented:
@imjasonh this applies post-merge, and so we will need to do some work to keep that from breaking.

If we go down that rabbit hole we should probably move this out of `mattmoor/` 😅

imjasonh commented:
> @imjasonh this applies post-merge, and so we will need to do some work to keep that from breaking.
> 
> If we go down that rabbit hole we should probably move this out of `mattmoor/` 😅

I split it into modules so you can apply only the image if you want, and I updated the workflows to do that instead as a test.

If we go further I'd love to have this auto-update on a schedule from GitHub, so I never have to think about it. If that's not your dream too, I'm happy to fork this 😄

mattmoor commented:
@imjasonh I think making a template that includes auto-updates would be great, I'm just reluctant to do it to my personal workstation stuff with the useless minimal image 🤣 

I can repo transfer this to wolfi-dev or chainguard-dev and we can get that set up when you are back?

imjasonh commented:
https://github.com/imjasonh/wolfstation/compare/deploy...imjasonh:wolfstation:setup sketches out the one-time setup to let GH auto-deploy workstation image updates.
